README.rst
checkout
------------

|build status|_

.. |build status| image:: https://secure.travis-ci.org/tuomasb/python-checkout.png?branch=master
   :alt: Build Status
.. _build status: http://travis-ci.org/tuomasb/python-checkout

Python wrapper for the Checkout Finland API.

Installation
------------

::

    $ pip install checkout

Resources
---------

- `Issue Tracker <http://github.com/tuomasb/python-checkout/issues>`_
- `Code <http://github.com/tuomasb/python-checkout>`_
- `Development Version
  <http://github.com/tuomasb/python-checkout/zipball/master#egg=checkout-dev>`_


CHANGES.rst
Changelog
---------

Here you can see the full list of changes between each python-checkout
release.

0.3.0 (February 6, 2014)
^^^^^^^^^^^^^^^^^^^^

- Initual public release of python-checkout.

0.2.0 (June 6, 2013)
^^^^^^^^^^^^^^^^^^^^

- Initial public release of python-verkkomaksut.


test_checkout.py
# -*- coding: utf-8 -*-
"""
    test_checkout
    ~~~~~~~~~~~~~~~~~

    Tests for `checkout` module.

    :copyright: (c) 2012 by Janne Vanhala.
    :license: BSD, see LICENSE for more details.
"""
import requests
from flexmock import flexmock
from pytest import raises
from checkout import (
    Checkout,
    Contact,
    Payment,
    CheckoutException
)


class TestCheckoutException(object):
    def setup_method(self, method):
        self.exception = CheckoutException('message')

    def test_is_an_exception(self):
        assert isinstance(self.exception, Exception)

    def test_sets_error_message(self):
        assert self.exception.message == 'message'


class TestContactWithSomeParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            address='Esimerkkikatu 123',
            postcode='01234',
            postoffice='Helsinki',
            country='FIN'
        )

    def test_sets_first_name(self):
        assert self.contact.first_name == 'Matti'

    def test_sets_last_name(self):
        assert self.contact.last_name == 'Meikäläinen'

    def test_sets_address(self):
        assert self.contact.address == 'Esimerkkikatu 123'

    def test_sets_postal_code(self):
        assert self.contact.postcode == '01234'

    def test_sets_postal_office(self):
        assert self.contact.postoffice == 'Helsinki'

    def test_sets_country(self):
        assert self.contact.country == 'FIN'

    def test_phone_is_blank_by_default(self):
        assert self.contact.phone == ''

    def test_email_is_blank_by_default(self):
        assert self.contact.email == ''


class TestContactWithAllParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            email='matti.meikalainen@gmail.com',
            address='Esimerkkikatu 123',
            postcode='01234',
            postoffice='Helsinki',
            country='FIN',
            phone='020123456',
        )

    def test_sets_phone(self):
        assert self.contact.phone == '020123456'

    def test_sets_email(self):
        assert self.contact.email == 'matti.meikalainen@gmail.com'

    def test_dict(self):
        assert self.contact.dict == {
            'PHONE': '020123456',
            'EMAIL': 'matti.meikalainen@gmail.com',
            'FIRSTNAME': 'Matti',
            'FAMILYNAME': 'Meikäläinen',
            'ADDRESS': 'Esimerkkikatu 123',
            'POSTCODE': '01234',
            'POSTOFFICE': 'Helsinki',
            'COUNTRY': 'FIN'
        }


class TestPaymentWithMinimumParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            address='Esimerkkikatu 123',
            postcode='01234',
            postoffice='Helsinki',
            country='FIN'
        )
        self.payment = Payment(
            order_number='12345678',
            reference_number='9999999',
            amount='200',
            delivery_date='20140606',
            return_url='https://www.esimerkkikauppa.fi/sv/return',
            cancel_url='https://www.esimerkkikauppa.fi/sv/cancel',
            contact=self.contact
        )

    def test_sets_order_number(self):
        assert self.payment.order_number == '12345678'

    def test_sets_reference_number(self):
        assert self.payment.reference_number == '9999999'

    def test_sets_amount(self):
        assert self.payment.amount == '200'

    def test_sets_delivery_date(self):
        assert self.payment.delivery_date == '20140606'

    def test_sets_return_url(self):
        assert self.payment.return_url == \
            'https://www.esimerkkikauppa.fi/sv/return'

    def test_sets_cancel_url(self):
        assert self.payment.cancel_url == \
            'https://www.esimerkkikauppa.fi/sv/cancel'

    def test_sets_contact(self):
        assert self.payment.contact is self.contact

    def test_message_is_blank_by_default(self):
        assert self.payment.message == ''

    def test_currency_is_euro_by_default(self):
        assert self.payment.currency == 'EUR'

    def test_language_is_finnish_by_default(self):
        assert self.payment.language == 'FI'

    def test_delayed_url_is_blank_by_default(self):
        assert self.payment.delayed_url == ''

    def test_reject_url_is_blank_by_default(self):
        assert self.payment.reject_url == ''

    def test_unsupported_currency_raises_value_error(self):
        with raises(CheckoutException) as exc_info:
            self.payment.currency = 'USD'
        assert exc_info.value.message == \
            'Currently EUR is the only supported currency.'

    def test_language_can_be_finnish(self):
        self.payment.language = 'FI'
        assert self.payment.language == 'FI'

    def test_language_can_be_swedish(self):
        self.payment.language = 'SE'
        assert self.payment.language == 'SE'

    def test_language_can_be_english(self):
        self.payment.language = 'EN'
        assert self.payment.language == 'EN'

    def test_unsupported_locale_raises_value_error(self):
        with raises(CheckoutException) as exc_info:
            self.payment.language = 'DE'
        assert exc_info.value.message == \
            "Given language is not supported: 'DE'"

    def test_dict(self):
        assert self.payment.dict == {
            'VERSION': '0001',
            'STAMP': '12345678',
            'AMOUNT': '200',
            'REFERENCE': '9999999',
            'MESSAGE': '',
            'LANGUAGE': 'FI',
            'RETURN': 'https://www.esimerkkikauppa.fi/sv/return',
            'CANCEL': 'https://www.esimerkkikauppa.fi/sv/cancel',
            'REJECT': '',
            'DELAYED': '',
            'CURRENCY': 'EUR',
            'CONTENT': '1',
            'TYPE': '0',
            'ALGORITHM': '3',
            'DELIVERY_DATE': '20140606',
            'PHONE': '',
            'EMAIL': '',
            'FIRSTNAME': 'Matti',
            'FAMILYNAME': 'Meikäläinen',
            'ADDRESS': 'Esimerkkikatu 123',
            'POSTCODE': '01234',
            'POSTOFFICE': 'Helsinki',
            'COUNTRY': 'FIN'
        }

class TestPaymentWithAllParameters(object):
    def setup_method(self, method):
        self.contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            email='matti.meikalainen@gmail.com',
            address='Esimerkkikatu 123',
            postcode='01234',
            postoffice='Helsinki',
            country='FIN',
            phone='020123456',
        )
        self.payment = Payment(
            order_number='12345678',
            reference_number='9999999',
            amount='200',
            delivery_date='20140606',
            message='Esimerkkimaksun kuvaus',
            currency='EUR',
            language='FI',
            content='10',
            return_url='https://www.esimerkkikauppa.fi/sv/return',
            cancel_url='https://www.esimerkkikauppa.fi/sv/cancel',
            delayed_url='https://www.esimerkkikauppa.fi/sv/delayed',
            reject_url='https://www.esimerkkikauppa.fi/sv/reject',
            contact=self.contact
        )

    def test_sets_message(self):
        assert self.payment.message == 'Esimerkkimaksun kuvaus'

    def test_sets_currency(self):
        assert self.payment.currency == 'EUR'

    def test_sets_language(self):
        assert self.payment.language == 'FI'

    def test_sets_content(self):
        assert self.payment.content == '10'

    def test_sets_delayed_url(self):
        assert self.payment.delayed_url == \
            'https://www.esimerkkikauppa.fi/sv/delayed'

    def test_sets_reject_url(self):
        assert self.payment.reject_url == \
            'https://www.esimerkkikauppa.fi/sv/reject'

    def test_dict(self):
        assert self.payment.dict == {
            'VERSION': '0001',
            'STAMP': '12345678',
            'AMOUNT': '200',
            'REFERENCE': '9999999',
            'MESSAGE': 'Esimerkkimaksun kuvaus',
            'LANGUAGE': 'FI',
            'RETURN': 'https://www.esimerkkikauppa.fi/sv/return',
            'CANCEL': 'https://www.esimerkkikauppa.fi/sv/cancel',
            'REJECT': 'https://www.esimerkkikauppa.fi/sv/reject',
            'DELAYED': 'https://www.esimerkkikauppa.fi/sv/delayed',
            'CURRENCY': 'EUR',
            'CONTENT': '10',
            'TYPE': '0',
            'ALGORITHM': '3',
            'DELIVERY_DATE': '20140606',
            'PHONE': '020123456',
            'EMAIL': 'matti.meikalainen@gmail.com',
            'FIRSTNAME': 'Matti',
            'FAMILYNAME': 'Meikäläinen',
            'ADDRESS': 'Esimerkkikatu 123',
            'POSTCODE': '01234',
            'POSTOFFICE': 'Helsinki',
            'COUNTRY': 'FIN'
        }

class TestPaymentWithoutContact(object):
    def setup_method(self, method):
        self.payment = Payment(
            order_number='12345678',
            reference_number='9999999',
            amount='200',
            delivery_date='20140606',
            message='Esimerkkimaksun kuvaus',
            currency='EUR',
            language='FI',
            content='10',
            return_url='https://www.esimerkkikauppa.fi/sv/return',
            cancel_url='https://www.esimerkkikauppa.fi/sv/cancel',
            delayed_url='https://www.esimerkkikauppa.fi/sv/delayed',
            reject_url='https://www.esimerkkikauppa.fi/sv/reject',
        )

    def test_dict(self):
        assert self.payment.dict == {
            'VERSION': '0001',
            'STAMP': '12345678',
            'AMOUNT': '200',
            'REFERENCE': '9999999',
            'MESSAGE': 'Esimerkkimaksun kuvaus',
            'LANGUAGE': 'FI',
            'RETURN': 'https://www.esimerkkikauppa.fi/sv/return',
            'CANCEL': 'https://www.esimerkkikauppa.fi/sv/cancel',
            'REJECT': 'https://www.esimerkkikauppa.fi/sv/reject',
            'DELAYED': 'https://www.esimerkkikauppa.fi/sv/delayed',
            'CURRENCY': 'EUR',
            'CONTENT': '10',
            'TYPE': '0',
            'ALGORITHM': '3',
            'DELIVERY_DATE': '20140606',
            'PHONE': '',
            'EMAIL': '',
            'FIRSTNAME': '',
            'FAMILYNAME': '',
            'ADDRESS': '',
            'POSTCODE': '',
            'POSTOFFICE': '',
            'COUNTRY': ''
        }

class TestCheckout(object):
    def test_defaults_to_merchant_test_account(self):
        checkout = Checkout()
        assert checkout.merchant_id == '375917'
        assert checkout.merchant_secret == 'SAIPPUAKAUPPIAS'

    def test_custom_merchant_credentials(self):
        checkout = Checkout(merchant_id='12345', merchant_secret='secret')
        assert checkout.merchant_id == '12345'
        assert checkout.merchant_secret == 'secret'

class TestReturnChecksum(object):
    def test_validate_payment_valid_return(self):
        checkout = Checkout()
        assert checkout.validate_payment_return(
            mac="2657BA96CC7879C79192547EB6C9D4082EA39CA52FE1DAD09CB1C632ECFDAE67",
            version="0001",
            order_number="1388998411",
            order_reference="474738238",
            payment="12288575",
            status="3",
            algorithm="3"
        )

    def test_validate_payment_invalid_return(self):
        checkout = Checkout()
        assert not checkout.validate_payment_return(
            mac="2657BA96CC7879C79192547EB6C9D4082EA39CA52FE1DAD09CB1C632ECFDAE68",
            version="0001",
            order_number="1388998411",
            order_reference="474738238",
            payment="1221238575",
            status="3",
            algorithm="3"
        )

class TestParameterLimitation(object):
    def test_exception_generation(self):
        with raises(CheckoutException) as exc_info:
            payment = Payment(
                order_number='12345678',
                reference_number='99999999999999999999999999999999999999999999',
                amount='200',
                delivery_date='20140606',
                message='Esimerkkimaksun kuvaus',
                currency='EUR',
                language='FI',
                content='10',
                return_url='https://www.esimerkkikauppa.fi/sv/return',
                cancel_url='https://www.esimerkkikauppa.fi/sv/cancel',
                delayed_url='https://www.esimerkkikauppa.fi/sv/delayed',
                reject_url='https://www.esimerkkikauppa.fi/sv/reject',
            )

        assert exc_info.value.message == \
            'reference_number over maximum allowed 20 characters'

    def test_parameter_clipping(self):
        contact = Contact(
            first_name='Matti',
            last_name='Meikäläinen',
            address='Esimerkkikatu 123',
            postcode='01234123123123123123123123123123123123123',
            postoffice='Helsinki',
            country='FIN'
        )
        assert len(contact.postcode) == 14


checkout/__init__.py
# -*- coding: utf-8 -*-
"""
    checkout
    ~~~~~~~~~~~~

    Python wrapper for the Checkout Finland API.

    :copyright: (c) 2013 by Janne Vanhala.
    :license: BSD, see LICENSE for more details.
"""
__version__ = '0.2.0'

import hashlib
import hmac
import json
import requests
import xml.etree.ElementTree as ET

class CheckoutException(Exception):
    """This exception is raised when the request made to the Checkout API
    is invalid, or some other error occurs in the usage of the API."""

    def __init__(self, message):
        #: An error description of the error in chosen localization. This error
        #: description is not meant to be displayed to the end-user.
        self.message = message


class Contact(object):
    """This class represents the payer of a payment.
    Contact details are optional """

    def __init__(self, **options):
        #: Payer's first name.
        self.first_name = options.get('first_name', '')[0:40]

        #: Payer's surname.
        self.last_name = options.get('last_name', '')[0:40]

        #: Payer's email address.
        self.email = options.get('email', '')[0:200]

        #: Payer's telephone number.
        self.phone = options.get('phone', '')[0:30]

        #: Payer's street address.
        self.address = options.get('address', '')[0:40]

        #: Payer's postal code.
        self.postcode = options.get('postcode', '')[0:14]

        #: Payer's post office.
        self.postoffice = options.get('postoffice', '')[0:18]

        #: Payer's country. 3-letter ISO code.
        self.country = options.get('country', '')[0:3]

    @property
    def dict(self):
        """Dict of this contact in fields specified by Checkout API and clipped accordingly."""
        return {
            'PHONE': self.phone,
            'EMAIL': self.email,
            'FIRSTNAME': self.first_name,
            'FAMILYNAME': self.last_name,
            'ADDRESS': self.address,
            'POSTCODE': self.postcode,
            'POSTOFFICE': self.postoffice,
            'COUNTRY': self.country
        }


class Payment(object):
    def __init__(self, order_number, reference_number, amount, delivery_date, return_url, cancel_url, **options):
        #: Order number is a string of characters identifying the customer's
        #: purchase and the used webshop software creates it. Mandatory.
        if len(order_number) > 20:
            raise CheckoutException("order_number over maximum allowed 20 characters")
        else:            
            self.order_number = order_number

        #: Reference number is sent to bank by default and is automatically
        #: created.  In those payment methods that are used as an interface,
        #: this field can contain own reference number, which is sent to the
        #: bank service instead of the automatically generated reference
        #: number. Mandatory.
        if len(reference_number) > 20:
            raise CheckoutException("reference_number over maximum allowed 20 characters")
        else:            
            self.reference_number = reference_number

        #: Order amount in cents. Mandatory.
        if len(amount) > 8:
            raise CheckoutException("amount over maximum allowed 8 characters")
        else:            
            self.amount = amount

        #: Delivery date of order in format YYYYMMDD. Mandatory
        if len(delivery_date) > 8:
            raise CheckoutException("delivery_date over maximum allowed 8 characters")
        else:            
            self.delivery_date = delivery_date

        #: Any data about the order in text format can be sent to the payment
        #: system. They are shown in the Merchant's Panel in payment details. Optional.
        self.message = options.get('message', '')[0:1000]

        #: Payment currency.  Value must EUR for the Finnish banks, otherwise
        #: the payment will not be accepted. Mandatory, defaults to 'EUR'.
        self.currency = options.get('currency', 'EUR')

        #: Language defines default language for the payment method
        #: selection page. Optional, 2-letter ISO code.
        self.language = options.get('language', 'FI')

        #: Contact object for the Payment. Optional, if supplied with None blank contact is used.
        self.contact = options.get('contact', Contact()) 

        #: Payment content. "1" for normal content and "10" for adult content. Mandatory, default 1.
        self.content = options.get('content', '1')[0:2]

        #: URL to which user is redirected after a successful payment. Mandatory.
        if len(return_url) > 300:
            raise CheckoutException("return_url over maximum allowed 300 characters")
        else:            
            self.return_url = return_url

        #: URL to which user is redirected after a cancelled or failed payment. Mandatory.
        if len(cancel_url) > 300:
            raise CheckoutException("cancel_url over maximum allowed 300 characters")
        else:            
            self.cancel_url = cancel_url

        #: URL to which user is directed, if the payment is pending.
        #: After the actual payment, the payment acknowledged as received by Checkout
        #: with fetching this URL along with same parameters as used in normal return_url.
        #: Optional.
        self.delayed_url = options.get('delayed_url', '')
        if len(self.delayed_url) > 300:
            raise CheckoutException("delayed_url over maximum allowed 300 characters")

        #: URL requested when the payment is marked as rejected.  The URL is
        #: requested with the same GET parameters as return address when the
        #: payment is made. Optional.
        self.reject_url = options.get('reject_url', '')
        if len(self.reject_url) > 300:
            raise CheckoutException("reject_url over maximum allowed 300 characters")

    @property
    def currency(self):
        return self._currency

    @currency.setter
    def currency(self, value):
        if value != 'EUR':
            raise CheckoutException("Currently EUR is the only supported currency.")
        self._currency = value

    @property
    def language(self):
        return self._language

    @language.setter
    def language(self, value):
        if value not in ('FI', 'SE', 'EN'):
            raise CheckoutException("Given language is not supported: %r" % value)
        self._language = value

    @property
    def dict(self):
        returndict = {
            'VERSION': "0001", #: Version of the API.
            'STAMP': self.order_number,
            'AMOUNT': self.amount,
            'REFERENCE': self.reference_number,
            'MESSAGE': self.message,
            'LANGUAGE': self.language,
            'RETURN': self.return_url,
            'CANCEL': self.cancel_url,
            'REJECT': self.reject_url,
            'DELAYED': self.delayed_url,
            'CURRENCY': self.currency,
            'CONTENT': self.content,
            'TYPE': "0", #: Static field.
            'ALGORITHM': "3", #: Return AUTHCODE algorithm, "3" for HMAC-SHA256.
            'DELIVERY_DATE': self.delivery_date
        }
        #: Merge with Contact values
        returndict.update(self.contact.dict)
        return returndict

class Checkout(object):
    SERVICE_URL = "https://payment.checkout.fi/"

    def __init__(self, merchant_id='375917',
                       merchant_secret='SAIPPUAKAUPPIAS'):
        """
        Initialize Checkout with your own merchant id and merchant secret.

        :param merchant_id: Mercant ID is given to you by Checkout
            when you make the contract. Default is the test merchant_id.
        :param merchant_secret: Merchant secret is given to you by Checkout.
            Default is the test merchant_secret.
        """
        self.merchant_id = merchant_id
        self.merchant_secret = merchant_secret
        self.session = requests.Session()

    def get_onsite_button_data(self, payment):
        """
        Creates a new payment and returns a `list` with the following data for each payment method:
        { 'bank': bankname, 'url': posturl, 'icon': iconurl, formfields: {} }

        :param payment: a `Payment` object
        """
        postdict = payment.dict
        postdict['MERCHANT'] = self.merchant_id
        postdict['DEVICE'] = "10" #: "10" to get XML data for payment methods back
        postdict['MAC'] = self._calculate_payment_md5(postdict, self.merchant_secret)
        response = self.session.post(self.SERVICE_URL, data=postdict)
        return self.parse_xml_response(response.content)

    def get_offsite_button_data(self, payment):
        """
        Returns form fields for off-page payment where user is sent to checkout.fi and shown
        all the payment options there instead of showing them onsite.

        :param payment: a `Payment` object
        """
        paymentdict = payment.dict
        paymentdict['MERCHANT'] = self.merchant_id
        paymentdict['DEVICE'] = "1" #: "1" to get payment method selection form from Checkout.fi
        paymentdict['MAC'] = self._calculate_payment_md5(paymentdict, self.merchant_secret)
        return paymentdict

    def parse_xml_response(self, xmlraw):
        """
        Parses XML-response for onsite payment method

        :param xmlraw: Raw XML data returned by checkout.fi
        """
        payment_list = []
        XML = ET.fromstring(xmlraw)
        banks = XML.findall(".//payment/banks/*")
        for bank in banks:
            bankdict = dict(bank.items())
            fielddict = {}
            for fieldname in bank:
                fielddict[fieldname.tag] = fieldname.text
            bankdict["fields"] = fielddict
            payment_list.append(bankdict)
        return payment_list

    def _calculate_payment_md5(self, params, merchant_secret):
        fields = [params["VERSION"], params["STAMP"], params["AMOUNT"], params["REFERENCE"], \
                  params["MESSAGE"], params["LANGUAGE"], params["MERCHANT"], params["RETURN"], \
                  params["CANCEL"], params["REJECT"], params["DELAYED"], params["COUNTRY"], \
                  params["CURRENCY"], params["DEVICE"], params["CONTENT"], params["TYPE"], \
                  params["ALGORITHM"], params["DELIVERY_DATE"], params["FIRSTNAME"], params["FAMILYNAME"], \
                  params["ADDRESS"], params["POSTCODE"], params["POSTOFFICE"], merchant_secret]

        base = '+'.join(fields)
        return hashlib.md5(base).hexdigest().upper()

    def validate_payment_return(self, mac, version, order_number, order_reference, payment, status, algorithm):
        """
        Validates parameters sent by Checkout Finland to the success/cancel URL or
        delayed/reject URL after a payment. The parameters must be validated
        in order to avoid hacking attempts to confirm payment. Returns `True`
        when the parameters are valid, and `False` otherwise.

        :param mac: A hash value calculated by payment system and sent to return url.
        :param version: Payment version number. GET parameter 'VERSION'.
        :param order_number: The same order number that was previously sent to
            the payment system. GET parameter 'STAMP'.
        :param order_reference: The same order reference that was previously sent
            to the payment system. GET parameter 'REFERENCE'.
        :param payment: A payment identified produced by Checkout Finland used
            for calculating the hash.
        :param status: Payment status code, which is part of payment
            confirmation. '2'/'5'/'6'/'8'/'9'/'10' = success. '3' = delayed. '-1' = cancelled.
            '7' = manual activation required. GET parameter 'STATUS'.
            Specified here: http://checkout.fi/uploads/sopimukset/Checkout_1_4_rajapinta_api-v1.7.pdf
        :param status: Payment return algorithm version. This library uses version 3.
            GET parameter 'ALGORITHM'.
        """
        fields = [version, order_number, order_reference, payment, status, algorithm]
        base = '&'.join(fields)
        return mac == hmac.new(self.merchant_secret, base, hashlib.sha256).hexdigest().upper()
